File: problem-solving/_lc-3-medium-longest-substring-without-repeating-characters/2-quadratic.py
"""
https://leetcode.com/problems/longest-substring-without-repeating-characters

Complexity
----------
Time: O(n^2)
Space: O(n)
"""

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        n = len(s)
        mx = 0
        for i in range(n):
            seen = set()
            for j in range(i, n):
                seen.add(s[j])
                if len(seen) < j - i + 1:
                    break
                mx = max(mx, len(seen))
        return mx

File: problem-solving/_lc-3-medium-longest-substring-without-repeating-characters/2-quadratic.go
"""
https://leetcode.com/problems/longest-substring-without-repeating-characters

Complexity
----------
Time: O(n^2)
Space: O(n)
"""

type void struct{}

func max(x, y int) int {
    if x > y {
        return x
    }
    return y
}

func lengthOfLongestSubstring(s string) int {
    var null void
    ans := 0
    n := len(s)
    for i := 0; i < n; i++ {
        set := make(map[byte]void)
        for j := i; j < n; j++ {
            set[s[j]] = null
            if len(set) < j - i + 1 {
                break
            }
            ans = max(ans, len(set))
        }
    }
    return ans
}
